tests/test_losses.py

```diff
@@ -124,6 +124,60 @@ def test_rigids_riemannian_loss_and_grad(self):
 
         # self.assertTrue(np.allclose(result_2, curved_expected_2))
 
+    def test_caffe_gradient_checker(self):
+
+        print('\n #####  GRADIENT CHECKER (CAFFE) ##### \n')
+
+        print('Setting Test Parameters: ') # Adjustable Parameters
+        dx = 0.01
+        threshold = 0.01
+        y_pred = np.array([-0.662919 , -1.29015 , 1.30245 , 1.25567 , 1.09228 , 0.153399 ]) #Regularized
+        y_true = np.array([0.0349858 , 0.562573 , -1.94181 , -0.173889 , 0.127527 , -0.19467 ]) #Regularized
+
+        print('Stepsize: ', str(dx))
+        print('threshold: ', str(threshold))
+        print('y_pred', str(y_pred))
+        print('y_true', str(y_true))
+
+        print('\nTesting Baseline forward and gradient: ')
+
+        forward_loss = losses.rigids_riemannian_loss(y_pred, y_true)
+        print('Forward Loss: ', str(forward_loss))
+        
+        backward_grad = losses.rigids_riemannian_grad(y_pred, y_true)
+        print('Backward Grad: ', str(backward_grad))
+
+
+        for xdim in range(0,6):
+            print('\n ----- Testing dL/dy_pred[',str(xdim),'] ----- ')
+
+            dx_vec = np.zeros(6)
+            dx_vec[xdim] = dx
+
+            y_pred_plus_dx = y_pred + dx_vec
+            print('y_pred[',str(xdim),']+dx: ',str(y_pred_plus_dx))
+            y_pred_minus_dx = y_pred - dx_vec
+            print('y_pred[',str(xdim),']-dx: ',str(y_pred_minus_dx))
+            
+            forward_loss_y_pred_plus_dx = losses.rigids_riemannian_loss(y_pred_plus_dx, y_true)
+            print('Forward with +dx: ',str(forward_loss_y_pred_plus_dx))
+            forward_loss_y_pred_minus_dx = losses.rigids_riemannian_loss(y_pred_minus_dx, y_true)
+            print('Forward with -dx: ',str(forward_loss_y_pred_minus_dx))
+            
+            dy_dx = (forward_loss_y_pred_plus_dx - forward_loss_y_pred_minus_dx) / (dx*2)
+            print('Computed grad backward_grad[',str(xdim),']: ',str(backward_grad[xdim]))
+            print('Calculated grad dL/dy_pred[',str(xdim),']: ',str(dy_dx))
+            
+            diff = np.abs(backward_grad[xdim]-dy_dx)
+            scale = np.max(np.fabs(np.array([backward_grad[xdim],dy_dx,1.0])))
+
+            if diff > (scale * threshold):
```

Comment:
use assert statements, not if statements

Something like self.assertAlmostEqual or something

Reply:
there are 6 partials to check in total, wouldn't it be better to see which partial(s) it failed on instead of fixing them in order of assertion?

Reply:
just add multiple assertions for all your prior invariants :)